### DISCUSSION TITLE
Regarding Usage of global Mutex locks

### DISCUSSION THREAD

**Discussion by pstarunuce** [hmlongco/Factory#235] [General]
we have seen every Container has its own Container Manager and its having its own registrations. Any specifc reason of using a global mutex lock instead of having a lock at a container level.  Can Global mutex lock  create any kind of issues if we extensively in an app which has many  containers.

**Comment by hmlongco** [comment]
Some of Factory's debugging and testing functionality depends on accessing shared global variables for logging, tracing, and so on. As such, container-level locks wouldn't provide the proper level of protection should resolutions ever occur across multiple threads simultaneously.

Note that almost all of that code is conditionally compiled and included *solely* for debugging and testing builds. Those globals don't exist in production code.

Should also note that the global mutex is a recursive lock, which allows resolutions that depend on resolutions that depend on resolutions.

So no. Factory is thread safe, and multiple containers aren't a problem.
